File: src/gameplay/level-up/PassiveItemBase.ts
import { PassiveItemConfig, StatModifier } from '../../data/types';
import { IPassiveItem } from '../../core/interfaces/IPassiveItem';

export class PassiveItemBase implements IPassiveItem {
  readonly id: string;
  readonly config: PassiveItemConfig;
  readonly maxLevel: number;
  level: number = 1;

  constructor(config: PassiveItemConfig) {
    this.id = config.itemId;
    this.config = config;
    this.maxLevel = config.maxLevel;
  }

  getModifier(): StatModifier {
    const ld = this.config.levelData[this.level - 1];
    // Return the first modifier (primary stat)
    return ld.statModifiers[0];
  }

  getAllModifiers(): StatModifier[] {
    const ld = this.config.levelData[this.level - 1];
    return ld.statModifiers;
  }

  levelUp(): void {
    if (this.level < this.maxLevel) {
      this.level++;
    }
  }
}

File: Assets/_Project/Scripts/Gameplay/LevelUp/PassiveItemBase.cs
using UnityEngine;
using VampireSurvivors.Core;

namespace VampireSurvivors.Gameplay
{
    /// <summary>
    /// 被動道具基底類別 — 實作 IPassiveItem，持有設定檔並依等級回傳屬性修正。
    /// </summary>
    public class PassiveItemBase : IPassiveItem
    {
        /// <summary>被動道具設定檔。</summary>
        private readonly PassiveItemConfigSO config;

        /// <summary>當前等級（1 起始）。</summary>
        private int level = 1;

        /// <inheritdoc/>
        public PassiveItemConfigSO Config => config;

        /// <inheritdoc/>
        public int Level => level;

        /// <summary>
        /// 建構被動道具實例。
        /// </summary>
        /// <param name="config">被動道具設定檔。</param>
        public PassiveItemBase(PassiveItemConfigSO config)
        {
            this.config = config;
        }

        /// <inheritdoc/>
        public void LevelUp()
        {
            if (level < config.maxLevel)
                level++;
        }

        /// <inheritdoc/>
        public StatModifier GetModifier()
        {
            if (config.levelModifiers == null || config.levelModifiers.Length == 0)
                return StatModifier.Default;

            int index = Mathf.Clamp(level - 1, 0, config.levelModifiers.Length - 1);
            return config.levelModifiers[index];
        }
    }
}
